crates/wast/src/ast/types.rs

```diff
@@ -104,13 +72,68 @@ impl<'a> Parse<'a> for ValType<'a> {
             })
         } else if l.peek::<kw::exnref>() {
             parser.parse::<kw::exnref>()?;
-            Ok(ValType::Exnref)
+            Ok(ValType::Ref(RefType::Exn))
         } else if l.peek::<kw::eqref>() {
             parser.parse::<kw::eqref>()?;
-            Ok(ValType::Eqref)
+            Ok(ValType::Ref(RefType::Eq))
         } else if l.peek::<kw::i31ref>() {
             parser.parse::<kw::i31ref>()?;
-            Ok(ValType::I31ref)
+            Ok(ValType::Ref(RefType::I31))
+        } else {
+            Err(l.error())
+        }
+    }
+}
+
+/// The reference value types for a wasm module.
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+pub enum RefType<'a> {
+    Func,
+    Extern,
+    Exn,
+    Eq,
+    I31,
+    Type(ast::Index<'a>),
+    OptType(ast::Index<'a>),
+}
+
+impl<'a> From<TableElemType> for RefType<'a> {
+    fn from(elem: TableElemType) -> Self {
+        match elem {
+            TableElemType::Funcref => RefType::Func,
+            TableElemType::Externref => RefType::Extern,
+            TableElemType::Exnref => RefType::Exn,
+        }
+    }
+}
+
+impl<'a> Parse<'a> for RefType<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let mut l = parser.lookahead1();
+        if l.peek::<kw::func>() {
+            parser.parse::<kw::func>()?;
+            Ok(RefType::Func)
+        } else if l.peek::<kw::r#extern>() {
+            parser.parse::<kw::r#extern>()?;
+            Ok(RefType::Extern)
+        } else if l.peek::<kw::any>() {
+            parser.parse::<kw::any>()?;
```

Comment:
Are there any old tests using `(ref any)` after everything is renamed?

Reply:
No, there shouldn't really be any. Might be some in wabt, but they'll update at some point. I've removed the alias for this form.